Extend index tables' partitioning settings to be able to restore them from a backup with the same partitioning as before

jepett0 opened the pull request:
https://github.com/ydb-platform/ydb/issues/4955

There are 3 major commits in this PR, whose purposes are the following:
- 90162f55be34706f7b08797aa3a0fa1703d29eca Extend `ydb_table.proto` to be able to specify partitioning settings of the `indexImplTable` and explicit partition boundaries (or uniform partition count) of the index table at the moment of its creation. There is also a functional test using GRPC calls that shows that it becomes possible to create a table with extended index partitioning settings.
- c96c24f11babd322026f7f5e9860ba7cdf8c6142 Changes in cpp SDK to be able to describe extended partitioning settings and explicit partition boundaries (if any) of secondary indexes of the table. This is necessary to be able to dump and restore the table with the extended index partitioning info.
- 49321ece51d08509414e40ff7ba089a0691d9078 Unit tests and functional tests of the feature.

Note: each major commit is preceded by a minor refactoring change. For the best reviewing experience, please review these changes separately.

jepett0 committed:
refactoring before tests

jepett0 committed:
Index tables' partitioning settings backup / restore tests

jepett0 committed:
refactoring before protobuf changes

jepett0 committed:
Extended index table partitioning settings via GRPC calls

jepett0 committed:
refactoring before SDK changes

jepett0 commented:
> ⚪ `2024-06-18 10:41:40 UTC` Pre-commit [check](https://github.com/ydb-platform/ydb/actions/runs/9553025785/job/26360935117) for [7fa8cdd](https://github.com/ydb-platform/ydb/commit/7fa8cdd6f70f3e24ad6732db8f170308417df81e) has started. ⚪ `2024-06-18 10:44:36 UTC` Build **linux-x86_64-release-asan** is running... 🟢 `2024-06-18 10:45:42 UTC` Build successful. ⚪ `2024-06-18 10:45:52 UTC` Tests are running... 🔴 `2024-06-18 12:44:31 UTC` Some tests failed, follow the links below.
> 
> [Test history](https://nebius.testmo.net/automation/runs/view/32454) | [Test log](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/test_logs/ya_test.log)
> 
> TESTS	PASSED	ERRORS	FAILED	SKIPPED	MUTED[?](https://github.com/ydb-platform/ydb/tree/main/.github/config/muted_ya.txt)
> [15445](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/summary/ya-test.html)	[15349](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/summary/ya-test.html#PASS)	0	[37](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/summary/ya-test.html#FAIL)	[48](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/summary/ya-test.html#SKIP)	[11](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9553025785/ya-x86-64-asan/summary/ya-test.html#MUTE)

There is a direct [leak](https://nebius.testmo.net/automation/runs/results/32454?group_by=automation_run_tests:status_id&group_id=3&test_id=727652396) in one of the added unit tests. However, this is not a new problem. It can be reproduced on a YDB cluster that uses `ydbd` binary that was built with `--sanitize=leak` flag by exporting to S3 and then importing back and stopping the server.

I propose to fix the leak in an other PR

jepett0 committed:
Extend index description in SDK

jepett0 commented:
> ⚪ `2024-06-19 11:38:02 UTC` Pre-commit [check](https://github.com/ydb-platform/ydb/actions/runs/9581298531/job/26417778273) for [60c6ccf](https://github.com/ydb-platform/ydb/commit/60c6ccfa510bfcc78909e5286369bac05b1e228e) has started. ⚪ `2024-06-19 11:40:49 UTC` Build **linux-x86_64-release-asan** is running... 🟢 `2024-06-19 12:10:48 UTC` Build successful. ⚪ `2024-06-19 12:10:59 UTC` Tests are running... 🔴 `2024-06-19 14:04:32 UTC` Some tests failed, follow the links below.
> 
> [Test history](https://nebius.testmo.net/automation/runs/view/32692) | [Test log](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/test_logs/ya_test.log)
> 
> TESTS	PASSED	ERRORS	FAILED	SKIPPED	MUTED[?](https://github.com/ydb-platform/ydb/tree/main/.github/config/muted_ya.txt)
> [15464](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/summary/ya-test.html)	[15370](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/summary/ya-test.html#PASS)	0	[30](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/summary/ya-test.html#FAIL)	[53](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/summary/ya-test.html#SKIP)	[11](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/9581298531/ya-x86-64-asan/summary/ya-test.html#MUTE)
> 🟡 `2024-06-19 14:05:12 UTC` merge: [60c6ccf](https://github.com/ydb-platform/ydb/commit/60c6ccfa510bfcc78909e5286369bac05b1e228e) ydbd size 5.7 GiB **+540.3 KiB +0.009%** vs build main: [ba35e13](https://github.com/ydb-platform/ydb/commit/ba35e13f57674ba305bb94876ebc4a922ac791d1)
> 
> Build size details

There is a direct [leak](https://nebius.testmo.net/automation/runs/results/32692?group_by=automation_run_tests:status_id&group_id=3&test_id=731245037) in one of the added unit tests. However, this is not a new problem. It can be reproduced on a YDB cluster that uses `ydbd` binary that was built with `--sanitize=leak` flag by exporting to S3 and then importing back and stopping the server.

I propose to fix the leak in an other PR

CyberROFL commented on ydb/public/sdk/cpp/client/ydb_table/table.cpp:
```diff
@@ -2243,31 +2259,67 @@ ui64 TIndexDescription::GetSizeBytes() const {
+
+void TGlobalIndexSettings::SerializeTo(Ydb::Table::GlobalIndexSettings& settings) const {
+    *settings.mutable_partitioning_settings() = PartitioningSettings.GetProto();
+    std::visit([&settings](auto&& partitions) {
+            using T = std::decay_t<decltype(partitions)>;
+            if constexpr (std::is_same_v<T, ui64>) {
+                settings.set_uniform_partitions(partitions);
+            } else if constexpr (std::is_same_v<T, TExplicitPartitions>) {
+                partitions.SerializeTo(*settings.mutable_partition_at_keys());
+            }
+        }, Partitions
```
На первый взгляд выглядит как поехавшее форматирование. Лучше лямбу объявить отдельно.

ijon commented on ydb/core/protos/flat_scheme_op.proto:
```diff
@@ -989,6 +993,13 @@ message TIndexDescription {
+    // indexImplTable settings
+    oneof Partitions {
+        uint32 UniformPartitions = 10;
+        TExplicitPartitions PartitionAtKeys = 11;
```
`Keys` is misleading here. Split points/boundaries are merely prefixes (dividing planes in the key space) and most often are not precise enough to be a full sized keys.
And we are not bound to any compatibility issues here, this being a private proto spec.

jepett0 replied:
Changed:

```diff
     oneof Partitions {
         uint32 UniformPartitions = 10;
-        TExplicitPartitions PartitionAtKeys = 11;
+        TExplicitPartitions ExplicitPartitions = 11;
     }
```

```diff
 message TExplicitPartitions {
-    repeated TSplitBoundary SplitBoundary = 1;
+    repeated TSplitBoundary SplitBoundaries = 1;
 }
```
 (it seems to me that repeatable field names should be plural to better express the semantics of a collection)

jepett0 committed:
review fixes

- proto field naming
- more reasonable uniform partitions count and min partitions count settings in tests
- remove dead code

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_path_describer.cpp:
```diff
@@ -1222,6 +1222,13 @@ void TSchemeShard::DescribeTableIndex(const TPathId& pathId, const TString& name
     entry.SetDataSize(tableStats.DataSize + tableStats.IndexSize);
+
+    *entry.MutablePartitioningPolicy() = indexImplTable->PartitionConfig().GetPartitioningPolicy();
+    if (const auto& explicitPartitions = indexImplTable->TableDescription.GetSplitBoundary();
+        !explicitPartitions.empty()
+    ) {
```
Мой Visual Studio Code ругается, говорит, что не получается неявно конвертировать RepeatedPtrField в bool. Оставил как было

jepett0 commented on ydb/core/ydb_convert/table_description.cpp:
```diff
@@ -862,6 +862,10 @@ bool FillIndexDescription(NKikimrSchemeOp::TIndexedTableCreationConfig& out,
+        if (!FillIndexTablePartitioning(*indexDesc->MutableIndexImplTableDescription(), index, status, error)) {
+            return false;
+        }
+
```
I have deleted this comment entirely. The author of it is @CyberROFL, who is reviewing this PR too, so I suppose it is ok

jepett0 committed:
review fix

- repeated field naming

jepett0 committed:
remove S3 test

- They fail under leak sanitizer. Will investigate in scope of another ticket.

jepett0 commented:
> There is a direct [leak](https://nebius.testmo.net/automation/runs/results/32692?group_by=automation_run_tests:status_id&group_id=3&test_id=731245037) in one of the added unit tests. However, it is not a new problem. It can be reproduced on a YDB cluster that uses `ydbd` binary that was built with `--sanitize=leak` flag by exporting to S3 and then importing back and stopping the server.
> 
> I propose to fix the leak in an other PR. I have created a [ticket](https://github.com/ydb-platform/ydb/issues/5737) to track the issue.

I decided to remove this test from the PR. I will add it back after I fix the problem with the leak
